encoding/codecv7_types.go

```diff
@@ -0,0 +1,432 @@
+}
+
+// BlobDataProofForPointEvaluation computes the abi-encoded blob verification data.
+// Note: This method is not implemented for daBatchV7.
+func (b *daBatchV7) BlobDataProofForPointEvaluation() ([]byte, error) {
+	return nil, nil
+}
+
+// Blob returns the blob of the batch.
+func (b *daBatchV7) Blob() *kzg4844.Blob {
+	return b.blob
+}
+
+// BlobBytes returns the blob bytes of the batch.
+func (b *daBatchV7) BlobBytes() []byte {
+	return b.blobBytes
+}
+
+// MarshalJSON implements the custom JSON serialization for daBatchV7.
+// This method is designed to provide prover with batch info in snake_case format.
+func (b *daBatchV7) MarshalJSON() ([]byte, error) {
+	type daBatchV7JSON struct {
+		Version           CodecVersion `json:"version"`
+		BatchIndex        uint64       `json:"batch_index"`
+		BlobVersionedHash string       `json:"blob_versioned_hash"`
+		ParentBatchHash   string       `json:"parent_batch_hash"`
+	}
+
+	return json.Marshal(&daBatchV7JSON{
+		Version:           b.version,
+		BatchIndex:        b.batchIndex,
+		BlobVersionedHash: b.blobVersionedHash.Hex(),
+		ParentBatchHash:   b.parentBatchHash.Hex(),
+	})
+}
+
+// Version returns the version of the DABatch.
+func (b *daBatchV7) Version() CodecVersion {
+	return b.version
+}
+
+// SkippedL1MessageBitmap returns the skipped L1 message bitmap of the DABatch.
+// For daBatchV7, there is no skipped L1 message bitmap.
+func (b *daBatchV7) SkippedL1MessageBitmap() []byte {
+	return nil
+}
+
+// DataHash returns the data hash of the DABatch.
+// For daBatchV7, there is no data hash.
+func (b *daBatchV7) DataHash() common.Hash {
+	return common.Hash{}
+}
+
+type blobPayloadV7 struct {
+	initialL1MessageIndex     uint64
+	initialL1MessageQueueHash common.Hash
+	lastL1MessageQueueHash    common.Hash
+
+	// used for encoding
+	blocks []*Block
+
+	// used for decoding
+	daBlocks       []DABlock
+	l2Transactions []types.Transactions
+}
+
+func (b *blobPayloadV7) InitialL1MessageIndex() uint64 {
+	return b.initialL1MessageIndex
+}
+func (b *blobPayloadV7) InitialL1MessageQueueHash() common.Hash {
+	return b.initialL1MessageQueueHash
+}
+
+func (b *blobPayloadV7) LastL1MessageQueueHash() common.Hash {
+	return b.lastL1MessageQueueHash
+}
+
+func (b *blobPayloadV7) Blocks() []DABlock {
+	return b.daBlocks
+}
+
+func (b *blobPayloadV7) Transactions() []types.Transactions {
+	return b.l2Transactions
+}
+
+func (b *blobPayloadV7) Encode() ([]byte, error) {
+	payloadBytes := make([]byte, blobPayloadV7MinEncodedLength)
+
+	binary.BigEndian.PutUint64(payloadBytes[blobPayloadV7OffsetInitialL1MessageIndex:blobPayloadV7OffsetInitialL1MessageQueue], b.initialL1MessageIndex)
+	copy(payloadBytes[blobPayloadV7OffsetInitialL1MessageQueue:blobPayloadV7OffsetLastL1MessageQueue], b.initialL1MessageQueueHash[:])
+	copy(payloadBytes[blobPayloadV7OffsetLastL1MessageQueue:blobPayloadV7OffsetInitialL2BlockNumber], b.lastL1MessageQueueHash[:])
+
+	var initialL2BlockNumber uint64
+	if len(b.blocks) > 0 {
```

Comment:
Is an empty batch (0 blocks) valid? If yes then let's sure to add a test for it

Reply:
there's a check to not allow for empty batch creation via the codec which is the only way to publicly interact: https://github.com/scroll-tech/da-codec/blob/b71c047ffecc91720cd8110f9fac227c84ef3e29/encoding/codecv7.go#L75

However, I added this check more from a user's PoV that we not accidentally create empty batches. I think technically it should be okay to create empty batches. in that case the blob payload will only contain the metadata

Reply:
For `codecV7.NewDABatch(batch)`: Empty Batch, creation error=no blocks -> is tested [here](https://github.com/scroll-tech/da-codec/blob/cbed8b2507be70f82e6cb4663668c56ed0c72edf/encoding/codecv7_test.go#L120) and [here](https://github.com/scroll-tech/da-codec/blob/cbed8b2507be70f82e6cb4663668c56ed0c72edf/encoding/codecv7_test.go#L212)

Reply:
I agree, technically empty batch is OK but we can disallow it.